File: Contests/ICPC-Nacional-2023-2024/i/i.cpp
#include<bits/stdc++.h>
#define ll long long

using namespace std;
  
const ll M = 1e9+7;

ll vet['z'+1];
ll qnt[112345];
 
map<char, ll> ma;
string s;
ll n, a, b, final, aux, ans;

int main(int argc, char const *argv[]){
    
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    cin >> s >> n;
    
    a = n % M;
    b = (n-1) % M;
    final = (a * b) / 2;
    final %= M;
 
    memset(qnt, 0, sizeof(qnt));
    memset(vet, 0, sizeof(vet));

    for (int i = s.size()-1; i >= 0 ; i--){
        ma[s[i]]++;
        vet[s[i]]++;
        for (int c = 'a'; c < s[i]; c++)
            qnt[i] += vet[c];
    }
    
    aux = 0;
    ans = 0;
    for (int i = 0; i < s.size(); i++){
        for (auto c : ma) 
            if (c.first < s[i])
                aux += c.second;
        
        ans += qnt[i] % M;
    }
    final = ((aux % M) * final) % M;
    ans = (ans * a) % M;
    ans += final;
    ans %= M;
 
    cout << ans << endl;
    return 0;
}

File: Contests/ICPC-Nacional-2023-2024/i/i.py

s = str(input())
n = int(input())

qnt = [0] * 112345

m = {
    'a' : 0,
    'b' : 0,
    'c' : 0,
    'd' : 0,
    'e' : 0,
    'f' : 0,
    'g' : 0,
    'h' : 0,
    'i' : 0,
    'j' : 0,
    'k' : 0,
    'l' : 0,
    'm' : 0,
    'n' : 0,
    'o' : 0,
    'p' : 0,
    'q' : 0,
    'r' : 0,
    's' : 0,
    't' : 0,
    'u' : 0,
    'v' : 0,
    'w' : 0,
    'x' : 0,
    'y' : 0,
    'z' : 0
}

MOD = 10**9 + 7

for i in range(len(s)-1, -1, -1):
    m[s[i]] += 1
    for j in range(97, ord(s[i]), 1):
        qnt[i]+=m[chr(j)]

ans = 0
aux = 0
for i in range(len(s)):
    for k, v in m.items():
        if (k < s[i]):
            aux += v
    ans += (n * qnt[i])
            
ans += aux * (((n-1) * n) // 2)


print(int(ans) % MOD)